src/RadixDlt.NetworkGateway.PostgresIntegration/Queries/NonFungibleVaultContentsQuery.cs

```diff
@@ -0,0 +1,247 @@
+        bool IsDeleted,
+        long NonFungibleIdDefinitionId,
+        string NonFungibleId,
+        byte[]? Data);
+
+    public record struct QueryConfiguration(
+        int NonFungibleIdsPerVault,
+        bool DescendingOrder,
+        StateVersionIdCursor? Cursor,
+        long AtLedgerState);
+
+    public static Task<Dictionary<long, ResultVault>> Execute(
+        ReadOnlyDbContext dbContext,
+        IDapperWrapper dapperWrapper,
+        ICollection<long> vaultEntityIds,
+        QueryConfiguration configuration,
+        CancellationToken token = default)
+    {
+        return ExecuteNonFungibleVaultContentsQuery(dbContext, dapperWrapper, vaultEntityIds, configuration, token);
+    }
+
+    private static async Task<Dictionary<long, ResultVault>> ExecuteNonFungibleVaultContentsQuery(
+        ReadOnlyDbContext dbContext,
+        IDapperWrapper dapperWrapper,
+        ICollection<long> vaultEntityIds,
+        QueryConfiguration configuration,
+        CancellationToken token)
+    {
+        if (vaultEntityIds.Count > 1 && configuration.Cursor != null)
+        {
+            throw new InvalidOperationException("Can't use cursor if executing against multiple vault entities.");
+        }
+
+        var cd = dapperWrapper.CreateCommandDefinition(
+            @"WITH
+variables AS (
+    SELECT
+        unnest(@vaultEntityIds) AS vault_entity_id,
+        @scanLimit AS scan_limit,
+        @pageLimit AS page_limit,
+        TRUE AS ignore_deleted,
+        TRUE AS include_data,
+        @descendingOrder AS descending_order,
+        ROW(@vaultCursorStateVersion, @vaultCursorId) AS vault_cursor,
+        @atLedgerState AS at_ledger_state
+)
+SELECT
+    var.vault_entity_id,
+    CAST(balance.balance AS TEXT) AS vault_balance,
+
+    entries.definition_id,
+    entries.definition_from_state_version,
+    entries.definition_is_last_candidate,
+    entries.last_updated_at_state_version,
+    entries.is_deleted,
+    entries.non_fungible_id_definition_id,
+    nf_def.non_fungible_id,
+    nf_data.data
+    -- , nf_data.is_deleted -- TODO how it relates to the entries.is_deleted? is it even needed?
+FROM variables var
+INNER JOIN LATERAL (
+    SELECT
+        ed.id AS definition_id,
+        ed.non_fungible_id_definition_id,
+        ed.from_state_version AS definition_from_state_version,
+        ed.is_last_candidate AS definition_is_last_candidate,
+        eh.from_state_version AS last_updated_at_state_version,
+        eh.is_deleted
+    FROM (
+        SELECT
+            *,
+            row_number() over (ORDER BY from_state_version DESC, id DESC) = var.scan_limit AS is_last_candidate
+        FROM non_fungible_vault_entry_definition
+        WHERE
+            vault_entity_id = var.vault_entity_id
+          AND from_state_version <= var.at_ledger_state
+          AND CASE WHEN var.descending_order THEN (from_state_version, id) <= var.vault_cursor ELSE (from_state_version, id) >= var.vault_cursor END
```

Comment:
- Yes, those are indeed inclusive cursors, hence `+ 1` thrown here and there. This is a more robust implementation as it eliminates the potential "missing last page" issue. All the cursors are meant to follow the exact same implementation to reduce confusion.
- I'm not quite sure what do you mean by "increment the cursor"?
- The implementation must be tested with automated tests that is currently being written alongside this PR.

Reply:
I can't find the +1 for these queries is what I'm trying to say :). By "increment the cursor" I mean go from "last result = (X, Y)" to "next cursor inclusive = (X, Y + 1)".

In general, we can't avoid the "missing last page" issue with this formulation where we do bounded lookups against definition; and filter out results. So I don't think we need to optimise for it -- although it might be nice if we attempt to find an N+1st element, and only return a cursor if we don't reach the end of the definitions whilst attempting to find an N+1st element.